# Sanity checks

In [ ]:
# Imports

import sys
sys.path.append('../..') # So we can import our modules 

import os
import pandas as pd

## Frame extraction

### Pornography-800

In [ ]:
data_loc_800 = "/nas-ctm01/datasets/public/BIOMETRICS/pornography-database/data-processed/middle_20"
os.path.isfile(f"{data_loc_800}/data.csv")

In [ ]:
frames_800 = pd.read_csv(f"{data_loc_800}/data.csv")
frames_800["video"] = [frame_name.split("#")[0] for frame_name in frames_800["frame"]]
frames_800

In [ ]:
agg = {"video": "first", "label": "first", "frame": "size"}
videos_800 = frames_800[["video", "frame", "label"]]
videos_800 = videos_800.groupby("video").aggregate(agg).reset_index(drop=True)
videos_800 = videos_800.rename(columns={"frame": "frame_count"})
videos_800

In [ ]:
# Videos from which it was not possible to extract 20 frames
videos_800[videos_800["frame_count"] < 20]

In [ ]:
len(videos_800) # 7 videos with 0 frames extracted

In [ ]:
len(frames_800[frames_800["label"] == 0]) # Number of non-porn frames

In [ ]:
len(frames_800[frames_800["label"] == 1]) # Number of porn frames

### Pornography-2k

In [ ]:
data_loc_2k = "/nas-ctm01/datasets/public/BIOMETRICS/pornography-2k-db/data-processed/middle_20"
os.path.isfile(f"{data_loc_2k}/data.csv")

In [ ]:
frames_2k = pd.read_csv(f"{data_loc_2k}/data.csv")
frames_2k["video"] = [frame_name.split("#")[0] for frame_name in frames_2k["frame"]]
frames_2k

In [ ]:
agg = {"video": "first", "label": "first", "frame": "size"}
videos_2k = frames_2k[["video", "frame", "label"]]
videos_2k = videos_2k.groupby("video").aggregate(agg).reset_index(drop=True)
videos_2k = videos_2k.rename(columns={"frame": "frame_count"})
videos_2k

In [ ]:
# Videos from which it was not possible to extract 40 frames
videos_2k[videos_2k["frame_count"] < 40]

In [ ]:
# Videos from which it was extracted more than 40 frames
videos_2k[videos_2k["frame_count"] > 40]

In [ ]:
len(frames_2k[frames_2k["label"] == 0]) # Number of non-porn frames

In [ ]:
len(frames_2k[frames_2k["label"] == 1]) # Number of porn frames

## Model training and testing

In [47]:
# TODO: change variables if running remotely
DATA_LOC = "../../data"
MODEL_NAME = "resnet50"
model_path = f"../../models/model_{MODEL_NAME}_freeze_False_epochs_10_batch_32_optim_sgd_optimized_False.pth"
results_path = f"../../results/model_{MODEL_NAME}_freeze_False_epochs_10_batch_32_optim_sgd_optimized_False.csv"

In [48]:
df_results = pd.read_csv(results_path)
df_results

Frame  Target  Prediction  Confidence
0         vPorn000601#0.jpg       1           1         1.0
1         vPorn000601#1.jpg       1           1         1.0
2         vPorn000601#2.jpg       1           1         1.0
3         vPorn000601#3.jpg       1           1         1.0
4         vPorn000601#4.jpg       1           1         1.0
...                     ...     ...         ...         ...
7995  vNonPorn000033#15.jpg       0           0         1.0
7996  vNonPorn000033#16.jpg       0           0         1.0
7997  vNonPorn000033#17.jpg       0           0         1.0
7998  vNonPorn000033#18.jpg       0           0         1.0
7999  vNonPorn000033#19.jpg       0           0         1.0

[8000 rows x 4 columns]

In [49]:
print("Videos correctly classified as pornographic:", len(df_results[(df_results["Target"] == 1) & (df_results["Prediction"] == 1)]))
print("Videos incorrectly classified as pornographic", len(df_results[(df_results["Target"] == 0) & (df_results["Prediction"] == 1)]))
print("Videos correctly classified as non-pornographic", len(df_results[(df_results["Target"] == 0) & (df_results["Prediction"] == 0)]))
print("Videos incorrectly classified as pornographic", len(df_results[(df_results["Target"] == 1) & (df_results["Prediction"] == 0)]))

Videos correctly classified as pornographic: 3586
Videos incorrectly classified as pornographic 377
Videos correctly classified as non-pornographic 3603
Videos incorrectly classified as pornographic 434


In [50]:
df_results["Video"] = df_results["Frame"].str.split("#").str[0]
df_results

Frame  Target  Prediction  Confidence           Video
0         vPorn000601#0.jpg       1           1         1.0     vPorn000601
1         vPorn000601#1.jpg       1           1         1.0     vPorn000601
2         vPorn000601#2.jpg       1           1         1.0     vPorn000601
3         vPorn000601#3.jpg       1           1         1.0     vPorn000601
4         vPorn000601#4.jpg       1           1         1.0     vPorn000601
...                     ...     ...         ...         ...             ...
7995  vNonPorn000033#15.jpg       0           0         1.0  vNonPorn000033
7996  vNonPorn000033#16.jpg       0           0         1.0  vNonPorn000033
7997  vNonPorn000033#17.jpg       0           0         1.0  vNonPorn000033
7998  vNonPorn000033#18.jpg       0           0         1.0  vNonPorn000033
7999  vNonPorn000033#19.jpg       0           0         1.0  vNonPorn000033

[8000 rows x 5 columns]

In [51]:
df_results[df_results["Frame"] == "vPorn000885#17.jpg"]

Frame  Target  Prediction  Confidence        Video
5297  vPorn000885#17.jpg       1           0    0.938795  vPorn000885

In [52]:
def separate_frames(row):
    target = row["Target"]

    correct_frames, incorrect_frames, correct_confidences, incorrect_confidences = [], [], [], []

    for frame, prediction, confidence in zip(row['Frame'], row['Prediction'], row['Confidence']):
        if target == prediction: # Correct prediction
            correct_frames.append(frame)
            correct_confidences.append(confidence)
        else: # Incorrect prediction
            incorrect_frames.append(frame)
            incorrect_confidences.append(confidence)

    return pd.Series(
        [correct_frames, incorrect_frames, correct_confidences, incorrect_confidences], 
        index=['Correct_frames', 'Incorrect_frames', 'Correct_confidences', 'Incorrect_confidences']
    )

# Group by video and aggregate columns
grouped_df = df_results.groupby('Video').agg({
    'Target': 'first',
    'Prediction': lambda x: list(x),
    'Frame': lambda x: list(x),
    'Confidence': lambda x: list(x)
}).reset_index()

# Apply aggregation function
grouped_df[['Correct_frames', 'Incorrect_frames', 'Correct_confidences', 'Incorrect_confidences']] = grouped_df.apply(separate_frames, axis=1)

# Drop unnecessary columns
grouped_df.drop(["Frame", "Prediction", "Confidence"], axis=1, inplace=True)

grouped_df

Video  Target  \
0    vNonPorn000024       0   
1    vNonPorn000030       0   
2    vNonPorn000031       0   
3    vNonPorn000033       0   
4    vNonPorn000045       0   
..              ...     ...   
395     vPorn000967       1   
396     vPorn000979       1   
397     vPorn000982       1   
398     vPorn000988       1   
399     vPorn000991       1   

                                        Correct_frames  \
0    [vNonPorn000024#0.jpg, vNonPorn000024#1.jpg, v...   
1    [vNonPorn000030#0.jpg, vNonPorn000030#1.jpg, v...   
2                                                   []   
3    [vNonPorn000033#0.jpg, vNonPorn000033#1.jpg, v...   
4    [vNonPorn000045#0.jpg, vNonPorn000045#1.jpg, v...   
..                                                 ...   
395                                                 []   
396  [vPorn000979#0.jpg, vPorn000979#1.jpg, vPorn00...   
397  [vPorn000982#0.jpg, vPorn000982#1.jpg, vPorn00...   
398  [vPorn000988#0.jpg, vPorn000988#1.jpg, vPorn00...   
399  [vPorn000991#0.jpg, vPorn000991#1.jpg, vPorn00...   

                                      Incorrect_frames  \
0                                                   []   
1                                                   []   
2    [vNonPorn000031#0.jpg, vNonPorn000031#1.jpg, v...   
3                                                   []   
4                                                   []   
..                                                 ...   
395  [vPorn000967#0.jpg, vPorn000967#1.jpg, vPorn00...   
396                                                 []   
397                                                 []   
398                                                 []   
399                                                 []   

                                   Correct_confidences  \
0    [0.9999995, 0.9999989, 0.9999993, 0.9999994, 0...   
1    [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
2                                                   []   
3    [0.9999664, 0.9998373, 0.9999877, 0.9999677, 0...   
4    [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
..                                                 ...   
395                                                 []   
396  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
397  [0.9999666, 0.99999547, 0.9999995, 0.999997, 0...   
398  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
399  [0.9999275, 0.9997297, 0.99964964, 0.9991555, ...   

                                 Incorrect_confidences  
0                                                   []  
1                                                   []  
2    [0.99740297, 0.99701476, 0.99594986, 0.9975376...  
3                                                   []  
4                                                   []  
..                                                 ...  
395  [0.9999875, 0.999992, 0.999998, 0.9999981, 0.9...  
396                                                 []  
397                                                 []  
398                                                 []  
399                                                 []  

[400 rows x 6 columns]

In [53]:
# Videos with both correctly and incorrectly classified frames
grouped_df[grouped_df.apply(lambda row: len(row["Correct_frames"]) > 0 and len(row["Incorrect_frames"]) > 0, axis=1)]

Video  Target  \
14   vNonPorn000074       0   
24   vNonPorn000125       0   
41   vNonPorn000221       0   
50   vNonPorn000262       0   
58   vNonPorn000301       0   
61   vNonPorn000307       0   
72   vNonPorn000354       0   
79   vNonPorn000383       0   
81   vNonPorn000394       0   
88   vNonPorn000430       0   
96   vNonPorn000472       0   
108  vNonPorn000528       0   
117  vNonPorn000555       0   
118  vNonPorn000556       0   
124  vNonPorn000584       0   
125  vNonPorn000585       0   
173  vNonPorn000858       0   
188  vNonPorn000939       0   
190  vNonPorn000950       0   
200     vPorn000028       1   
204     vPorn000054       1   
209     vPorn000081       1   
211     vPorn000085       1   
215     vPorn000104       1   
226     vPorn000180       1   
233     vPorn000224       1   
234     vPorn000226       1   
240     vPorn000260       1   
253     vPorn000324       1   
255     vPorn000334       1   
258     vPorn000351       1   
267     vPorn000415       1   
272     vPorn000434       1   
276     vPorn000454       1   
281     vPorn000474       1   
282     vPorn000475       1   
330     vPorn000674       1   
350     vPorn000757       1   
351     vPorn000767       1   
359     vPorn000811       1   
373     vPorn000885       1   

                                        Correct_frames  \
14   [vNonPorn000074#0.jpg, vNonPorn000074#1.jpg, v...   
24   [vNonPorn000125#1.jpg, vNonPorn000125#5.jpg, v...   
41   [vNonPorn000221#0.jpg, vNonPorn000221#1.jpg, v...   
50   [vNonPorn000262#0.jpg, vNonPorn000262#1.jpg, v...   
58   [vNonPorn000301#0.jpg, vNonPorn000301#1.jpg, v...   
61   [vNonPorn000307#0.jpg, vNonPorn000307#1.jpg, v...   
72   [vNonPorn000354#1.jpg, vNonPorn000354#2.jpg, v...   
79   [vNonPorn000383#0.jpg, vNonPorn000383#1.jpg, v...   
81                             [vNonPorn000394#13.jpg]   
88   [vNonPorn000430#1.jpg, vNonPorn000430#2.jpg, v...   
96   [vNonPorn000472#0.jpg, vNonPorn000472#1.jpg, v...   
108  [vNonPorn000528#0.jpg, vNonPorn000528#1.jpg, v...   
117  [vNonPorn000555#0.jpg, vNonPorn000555#1.jpg, v...   
118  [vNonPorn000556#0.jpg, vNonPorn000556#2.jpg, v...   
124  [vNonPorn000584#11.jpg, vNonPorn000584#12.jpg,...   
125  [vNonPorn000585#0.jpg, vNonPorn000585#1.jpg, v...   
173  [vNonPorn000858#0.jpg, vNonPorn000858#1.jpg, v...   
188  [vNonPorn000939#0.jpg, vNonPorn000939#1.jpg, v...   
190  [vNonPorn000950#3.jpg, vNonPorn000950#4.jpg, v...   
200  [vPorn000028#0.jpg, vPorn000028#1.jpg, vPorn00...   
204  [vPorn000054#0.jpg, vPorn000054#1.jpg, vPorn00...   
209  [vPorn000081#1.jpg, vPorn000081#2.jpg, vPorn00...   
211  [vPorn000085#0.jpg, vPorn000085#1.jpg, vPorn00...   
215  [vPorn000104#0.jpg, vPorn000104#11.jpg, vPorn0...   
226  [vPorn000180#0.jpg, vPorn000180#1.jpg, vPorn00...   
233  [vPorn000224#1.jpg, vPorn000224#2.jpg, vPorn00...   
234  [vPorn000226#0.jpg, vPorn000226#1.jpg, vPorn00...   
240  [vPorn000260#0.jpg, vPorn000260#2.jpg, vPorn00...   
253  [vPorn000324#0.jpg, vPorn000324#4.jpg, vPorn00...   
255  [vPorn000334#4.jpg, vPorn000334#5.jpg, vPorn00...   
258  [vPorn000351#0.jpg, vPorn000351#1.jpg, vPorn00...   
267  [vPorn000415#0.jpg, vPorn000415#1.jpg, vPorn00...   
272  [vPorn000434#3.jpg, vPorn000434#4.jpg, vPorn00...   
276  [vPorn000454#1.jpg, vPorn000454#2.jpg, vPorn00...   
281  [vPorn000474#1.jpg, vPorn000474#5.jpg, vPorn00...   
282  [vPorn000475#0.jpg, vPorn000475#1.jpg, vPorn00...   
330  [vPorn000674#1.jpg, vPorn000674#2.jpg, vPorn00...   
350  [vPorn000757#0.jpg, vPorn000757#1.jpg, vPorn00...   
351  [vPorn000767#5.jpg, vPorn000767#6.jpg, vPorn00...   
359  [vPorn000811#2.jpg, vPorn000811#3.jpg, vPorn00...   
373  [vPorn000885#0.jpg, vPorn000885#1.jpg, vPorn00...   

                                      Incorrect_frames  \
14   [vNonPorn000074#4.jpg, vNonPorn000074#5.jpg, v...   
24   [vNonPorn000125#0.jpg, vNonPorn000125#2.jpg, v...   
41   [vNonPorn000221#5.jpg, vNonPorn000221#7.jpg, v...   
50   [vNonPorn000262#3.jpg, vNonPorn000262

In [54]:
# Videos with only incorrectly classified frames
grouped_df[grouped_df.apply(lambda row: len(row["Correct_frames"]) == 0 and len(row["Incorrect_frames"]) > 0, axis=1)]

Video  Target Correct_frames  \
2    vNonPorn000031       0             []   
55   vNonPorn000290       0             []   
59   vNonPorn000304       0             []   
67   vNonPorn000343       0             []   
68   vNonPorn000345       0             []   
70   vNonPorn000352       0             []   
75   vNonPorn000367       0             []   
82   vNonPorn000395       0             []   
166  vNonPorn000808       0             []   
192  vNonPorn000966       0             []   
207     vPorn000076       1             []   
264     vPorn000395       1             []   
266     vPorn000413       1             []   
274     vPorn000450       1             []   
304     vPorn000567       1             []   
318     vPorn000622       1             []   
320     vPorn000625       1             []   
339     vPorn000714       1             []   
343     vPorn000732       1             []   
354     vPorn000785       1             []   
366     vPorn000834       1             []   
378     vPorn000912       1             []   
387     vPorn000933       1             []   
389     vPorn000938       1             []   
395     vPorn000967       1             []   

                                      Incorrect_frames Correct_confidences  \
2    [vNonPorn000031#0.jpg, vNonPorn000031#1.jpg, v...                  []   
55   [vNonPorn000290#0.jpg, vNonPorn000290#1.jpg, v...                  []   
59   [vNonPorn000304#0.jpg, vNonPorn000304#1.jpg, v...                  []   
67   [vNonPorn000343#0.jpg, vNonPorn000343#1.jpg, v...                  []   
68   [vNonPorn000345#0.jpg, vNonPorn000345#1.jpg, v...                  []   
70   [vNonPorn000352#0.jpg, vNonPorn000352#1.jpg, v...                  []   
75   [vNonPorn000367#0.jpg, vNonPorn000367#1.jpg, v...                  []   
82   [vNonPorn000395#0.jpg, vNonPorn000395#1.jpg, v...                  []   
166  [vNonPorn000808#0.jpg, vNonPorn000808#1.jpg, v...                  []   
192  [vNonPorn000966#0.jpg, vNonPorn000966#1.jpg, v...                  []   
207  [vPorn000076#0.jpg, vPorn000076#1.jpg, vPorn00...                  []   
264  [vPorn000395#0.jpg, vPorn000395#1.jpg, vPorn00...                  []   
266  [vPorn000413#0.jpg, vPorn000413#1.jpg, vPorn00...                  []   
274  [vPorn000450#0.jpg, vPorn000450#1.jpg, vPorn00...                  []   
304  [vPorn000567#0.jpg, vPorn000567#1.jpg, vPorn00...                  []   
318  [vPorn000622#0.jpg, vPorn000622#1.jpg, vPorn00...                  []   
320  [vPorn000625#0.jpg, vPorn000625#1.jpg, vPorn00...                  []   
339  [vPorn000714#0.jpg, vPorn000714#1.jpg, vPorn00...                  []   
343  [vPorn000732#0.jpg, vPorn000732#1.jpg, vPorn00...                  []   
354  [vPorn000785#0.jpg, vPorn000785#1.jpg, vPorn00...                  []   
366  [vPorn000834#0.jpg, vPorn000834#1.jpg, vPorn00...                  []   
378  [vPorn000912#0.jpg, vPorn000912#1.jpg, vPorn00...                  []   
387  [vPorn000933#0.jpg, vPorn000933#1.jpg, vPorn00...                  []   
389  [vPorn000938#0.jpg, vPorn000938#1.jpg, vPorn00...                  []   
395  [vPorn000967#0.jpg, vPorn000967#1.jpg, vPorn00...                  []   

                                 Incorrect_confidences  
2    [0.99740297, 0.99701476, 0.99594986, 0.9975376...  
55   [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...  
59   [0.9977337, 0.99953353, 0.9994492, 0.9993524, ...  
67   [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...  
68   [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...  
70   [0.9854938, 0.9980704, 0.9997373, 0.9998957, 0...  
75   [0.9999995, 0.99999976, 0.99999964, 0.9999995,...  
82   [0.97342664, 0.96763927, 0.9733001, 0.980342, ...  
166  [0.9989249, 0.9986197, 0.9976025, 0.9969367, 0...  
192  [0.9997745, 0.9997638, 0.8995691, 0.8633119, 0...  
207  [0.99999344, 0.99988496, 0.999736, 0.9997851, ...  
264  [0.98926824, 0.92212105, 0.8636849, 0.92616355...  
266  [0.9998234, 0.9928281, 0.98569393, 0.96950

In [55]:
# Videos with only correctly classified frames
grouped_df[grouped_df.apply(lambda row: len(row["Incorrect_frames"]) == 0 and row["Target"] == 1, axis=1)]

Video  Target                                     Correct_frames  \
199  vPorn000005       1  [vPorn000005#0.jpg, vPorn000005#1.jpg, vPorn00...   
201  vPorn000034       1  [vPorn000034#0.jpg, vPorn000034#1.jpg, vPorn00...   
202  vPorn000037       1  [vPorn000037#0.jpg, vPorn000037#1.jpg, vPorn00...   
203  vPorn000042       1  [vPorn000042#0.jpg, vPorn000042#1.jpg, vPorn00...   
205  vPorn000055       1  [vPorn000055#0.jpg, vPorn000055#1.jpg, vPorn00...   
..           ...     ...                                                ...   
394  vPorn000966       1  [vPorn000966#0.jpg, vPorn000966#1.jpg, vPorn00...   
396  vPorn000979       1  [vPorn000979#0.jpg, vPorn000979#1.jpg, vPorn00...   
397  vPorn000982       1  [vPorn000982#0.jpg, vPorn000982#1.jpg, vPorn00...   
398  vPorn000988       1  [vPorn000988#0.jpg, vPorn000988#1.jpg, vPorn00...   
399  vPorn000991       1  [vPorn000991#0.jpg, vPorn000991#1.jpg, vPorn00...   

    Incorrect_frames                                Correct_confidences  \
199               []  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
201               []  [0.9853506, 0.9880782, 0.9928063, 0.9779012, 0...   
202               []  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
203               []  [1.0, 1.0, 0.9999999, 1.0, 1.0, 1.0, 1.0, 1.0,...   
205               []  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
..               ...                                                ...   
394               []  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
396               []  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
397               []  [0.9999666, 0.99999547, 0.9999995, 0.999997, 0...   
398               []  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...   
399               []  [0.9999275, 0.9997297, 0.99964964, 0.9991555, ...   

    Incorrect_confidences  
199                    []  
201                    []  
202                    []  
203                    []  
205                    []  
..                    ...  
394                    []  
396                    []  
397                    []  
398                    []  
399                    []  

[164 rows x 6 columns]

In [56]:
# Videos which would have been correctly classified if in a majority voting scheme
print(f"Videos with more correctly than incorrectly classified frames: {len(grouped_df[grouped_df.apply(lambda row: len(row['Correct_frames']) > len(row['Incorrect_frames']), axis=1)])} / {len(grouped_df)}")

Videos with more correctly than incorrectly classified frames: 358 / 400


In [59]:
df_results[df_results["Frame"] == "vPorn000885#17.jpg"]

Frame  Target  Prediction  Confidence        Video
5297  vPorn000885#17.jpg       1           0    0.938795  vPorn000885

- vPorn000885#17 (5297) - really interesting example because when predicting porn, the model looks at the logo, but to predict non-porn the model looks at the sexual organ
- vPorn000676#9 (1789), vPorn000687#14 (3434) - interesting to see how the model is paying attention to the logo
- vNonPorn000367#3 (2383) - when it classifies a non-porn as porn, it is usually paying attention to actually relevant stuff
- vNonPorn000808#0 (4100) - here it was paying attention to the hands (skin)
- vNonPorn000966#0 (1520) - here it payed attention two the pigs, it might have been because of the shape or the color, which is similar to that of skin
- vPorn000005#0 (4140), vPorn000034#6 (686) - here, for example, although there is a logo, it payed attention to the skin areas